firedrake/adjoint_utils/blocks/solving.py

```diff
@@ -700,8 +723,10 @@ def _ad_assign_map(self, form, solver):
                            firedrake.Cofunction)):
                 coeff_count = coeff.count()
                 if coeff_count in form_ad_count_map:
-                    assign_map[form_ad_count_map[coeff_count]] = \
-                        block_variable.saved_output
+                    if solver == Solver.HESSIAN:
+                        assign_map[form_ad_count_map[coeff_count]] = block_variable.tlm_value
+                    else:
+                        assign_map[form_ad_count_map[coeff_count]] = block_variable.saved_output
```

Comment:
For the hessian solve don't you need to update from `tlm_value` _as well as_ from `saved_output`, rather than _instead of_? Otherwise the adjoint Jacobian in the `_ad_solvers["hessian_lvs"]` won't be assembled around the correct state.